Initial support for York AC protocol

danielegobbetti opened the pull request:
As mentioned in https://github.com/crankyoldgit/IRremoteESP8266/issues/1882#issuecomment-1250326149 I'm trying to add support for the protocol used by my (pretty old) AC unit by York (AC model: MHH07P17 remote GRYLH2A).

Please consider that I have little to no experience with C++ and therefore have lots of questions, nevertheless I wanted to open the PR since basic decoding works. I am unable to figure out the checksum protocol used by this remote.

So to the questions I have:
- [x] The AC unit has a timer for on and off power AND a sleep functionality (auto power off after some time): should I use OnTime or OnTime**r** in the "Native representation of the message"?
- [x] the on/off timer base units are 10 minutes, the timer is sent as an offset (the base unit has no clock AFAIK). I noticed in the code that also the internal representation of the timer should be in units of 10 minutes, but when using "minsToString" the output is in minutes. Should I multiply by 10?
- [x] The AC unit supports Vertical Swing but only in a on/off fashion: I am unsure of the mapping and of the usage of addSwingVToString
- [x] It is not clear to me what the "sleep" supported feature (i.e. ``result.sleep`` in toCommon()) actually is: is it meant to be a "night/low noise mode" or rather "power the unit of after some time" mode?
- [x] The on/offTimer do not seem to have a corresponding field in the state_t struct. Should these be mapped to some combination of clock / sleep?


I am aware of having also left some "TODOs" in the code.

Looking forward to some feedback regarding the questions above (and in general about the code).

danielegobbetti committed:
Add support for York AC protocol

The protocol is known to almost every detail, checksum is calculated properly.
Some basic tests have been added.

Setting the sleep timer, scheduled power on, scheduled power off do not work
since there are some flags within byte 12 which is not correctly populated ATM.

crankyoldgit commented:
First off, thanks for creating a PR. It's a big effort, I know. So congrats for making it thus far.
Don't worry, we will walk you through it.

I'll answer some of your questions before I start looking at your code.

> * The AC unit has a timer for on and off power AND a sleep functionality (auto power off after some time): should I use OnTime or OnTime**r** in the "Native representation of the message"?

"On Timer" i.e. `kOnTimerStr`
https://github.com/crankyoldgit/IRremoteESP8266/blob/726b219a692b3aaf7aa8c41f9d2cf48f05231dc8/src/locale/defaults.h#L90-L95
https://github.com/crankyoldgit/IRremoteESP8266/blob/726b219a692b3aaf7aa8c41f9d2cf48f05231dc8/src/IRtext.cpp#L66

> * the on/off timer base units are 10 minutes, the timer is sent as an offset (the base unit has no clock AFAIK). I noticed in the code that also the internal representation of the timer should be in units of 10 minutes, but when using "minsToString" the output is in minutes. Should I multiply by 10?

Please use the unit of Minutes. i.e. `setFunctionName(nr_of_mins)` and then divide by 10 etc internally. Ditto for a `getFunctionName()`. i.e. Multiply by 10. etc. This way it keeps things consistent between protocols. There are several other protocols that have this sort of limitation/implementation too.

> * The AC unit supports Vertical Swing but only in a on/off fashion: I am unsure of the mapping and of the usage of addSwingVToString

Don't use `addSwingVToString` in this case, as it is just on or off, it is a `boolean`, so use `addBoolToString`
e.g.
https://github.com/crankyoldgit/IRremoteESP8266/blob/9228143207ab7e2968a42d9a92e246237af7bb83/src/ir_Neoclima.cpp#L523

crankyoldgit commented:
The automatic code linter found a number of issues, these should hopefully be easy to understand and fix:
```
src/IRrecv.h:866:  Redundant blank line at the end of a code block should be deleted.  [whitespace/blank_line] [3]
src/ir_York.cpp:55:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.cpp:5[9](https://github.com/crankyoldgit/IRremoteESP8266/actions/runs/3129077653/jobs/5077712679#step:6:10):  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.cpp:67:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_York.cpp:67:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.cpp:76:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.cpp:83:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.cpp:99:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_York.cpp:99:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.cpp:198:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.cpp:270:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:279:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:287:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:295:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:303:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:303:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.cpp:345:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:345:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:29:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.h:31:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.h:32:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.h:33:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:36:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:38:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:39:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.h:40:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:42:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:44:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:45:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_York.h:46:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:48:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:50:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_York.h:56:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

crankyoldgit commented on src/IRsend.h:
```diff
@@ -837,6 +837,12 @@ class IRsend {
+                    const uint16_t nbytes = kYorkStateLength,
+                    const uint16_t repeat = kNoRepeat);
+#endif  // SEND_YORK
+
```
Style NitPick: please remove the extra blank line you've added at the end here.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+
+
+
+
```
Style Nitpick: reduce the number of blank lines to 2 at most.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+
+#if SEND_YORK
+// Alternative >64bit function to send YORK messages
+// Function should be safe over 64 bits.
```
Remove these two lines. They are just a note to the programmer from the automatically generated code.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+#if SEND_YORK
+// Alternative >64bit function to send YORK messages
+// Function should be safe over 64 bits.
+/// Send a  formatted message.
```
Change to something like:
`/// Send a 17 Byte / 136 bit York A/C message.`

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// e.g.
+/// @code
+///   uint8_t data[kStateLength] = {0x10, 0x08, 0xE0, 0x40, 0x02, 0x10, 0x80, 0x10, 0x80, 0x12, 0x00, 0x00, 0x00, 0x00, 0x37, 0xA0, 0x37};
+/// @endcode
```
You can remove this, as
a) It's a note for the programmer generated from the automatic generator.
b) You've determined it is in LSB order.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+                data + pos, 17,  // Bytes
+                kFreq, false, kNoRepeat, kDutyDefault); //false == LSB
+    pos += 17;  // Adjust by how many bytes of data we sent
+  }
```
You can simplify this automatically generated code to just:
```c++
  sendGeneric(kHdrMark, kHdrSpace,
              kBitMark, kOneSpace,
              kBitMark, kZeroSpace,
              kBitMark, kDefaultMessageGap,
              data, kYorkStateLength,  // Bytes
              kFreq, false, repeat, kDutyDefault);  // false == LSB
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+const uint16_t kOneSpace = 1778;
+const uint16_t kZeroSpace = 579;
+const uint16_t kFreq = 38000;  // Hz. (Guessing the most common frequency.)
+const uint16_t kOverhead = 3;
```
Please change these constants to include the Protocol name.
e.g.
```c++
const uint16_t kYorkHdrMark = 4887;
const uint16_t kYorkBitMark = 612;
etc.
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+  if (strict && nbits != kYorkBits)
+    return false;
+
+  uint16_t pos = 0;
```
You can safely remove all references to `pos` in this function, it isn't used.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+  uint16_t used = 0;
+
+  used = matchGeneric(results->rawbuf + offset, results->state + pos,
+                      results->rawlen - offset, 136,
```
Change `138` to `nbits`

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+                      false); //LSB
+  if (used == 0) return false;  // We failed to find any data.
+  offset += used;  // Adjust for how much of the message we read.
+  pos += 17;  // Adjust by how many bytes of data we read
```
You can remove these two lines. They were automatically generated code that isn't required in this situation.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+    case kYorkHeat: return stdAc::opmode_t::kHeat;
+    case kYorkDry:  return stdAc::opmode_t::kDry;
+    case kYorkFan:  return stdAc::opmode_t::kFan;
+    default:           return stdAc::opmode_t::kAuto;
```
Style Nitpick: Please line up all the `return` statements.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+
+/// Set the speed of the fan.
+/// @param[in] speed The desired setting.
+/// @note The speed is locked to Low when in Dry mode and to auto when in auto mode.
```
Your comment here doesn't match the code.
You're locking the result to Low when it is in `kYorkFan` mode.

danielegobbetti replied:
Great catch! Thanks a lot. Further, there was indeed an additional constraint in Fan mode. Added both as comment and as code.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRYork::IRYork(const uint16_t pin, const bool inverted,
```
Please change it from `IRYork` to `IRYorkAc` or `IRYorkAC`. That's just to match the style of other protocols in this library.

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,123 @@
+    //byte 8
+    uint8_t Mode     :8;  // 1 heat, 2 cool, 3 dry, 4 fan, 8 auto
+    //byte 9
+    uint8_t Temp     :8;  // Degrees Celsius x4: Min temp 18C 0x48 Max temp 32C 0x80
```
If you change this to:
```c++
    // Byte 9
    uint8_t               :2;
    uint8_t Temp     :6;  // Degrees Celsius: Min temp 18C Max temp 32C
```
It should save you having to do the multiple & divide by `4` in your get/set functions.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+void IRYork::setTemp(const uint8_t degrees) {
+  uint8_t temp = std::max(kYorkMinTemp, degrees);
+  temp = std::min(kYorkMaxTemp, temp);
+  _.Temp = (temp * 4);
```
See comment in the `ir_York.h` code.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// Get the current temperature setting.
+/// @return Get current setting for temp. in degrees celsius.
+uint8_t IRYork::getTemp(void) const {
+  return _.Temp / 4;
```
See comment in the `ir_York.h` code.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// @note Setting the On Timer active will cancel the Sleep timer/setting.
+void IRYork::setOnTimer(const uint16_t nr_of_mins) {
+  // TODO: York AC unit uses 10 minutes increments from now
+  _.OnTime = nr_of_mins;
```
Per answer to your question. use:
```c++
  _.OnTime = nr_of_mins / 10;
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// @note Setting the Off Timer active will cancel the Sleep timer/setting.
+void IRYork::setOffTimer(const uint16_t nr_of_mins) {
+  // TODO: York AC unit uses 10 minutes increments from now
+  _.OffTime = nr_of_mins;
```
Ditto.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// @return The Nr. of minutes the On Timer is set for.
+uint16_t IRYork::getOnTimer(void) const {
+  // TODO: York AC unit uses 10 minutes increments from now
+  return _.OnTime;
```
Per answer to your question. use:
```c++
  return _.OnTime * 10;
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+/// @note Sleep & Off Timer share the same timer.
+uint16_t IRYork::getOffTimer(void) const {
+  // TODO: York AC unit uses 10 minutes increments from now
+  return _.OffTime;
```
Ditto.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+  result.fanspeed = toCommonFanSpeed(_.Fan);
+  result.swingv = _.SwingV ? stdAc::swingv_t::kAuto : stdAc::swingv_t::kOff;
+  // Not supported.
+  result.model = -1;
+  result.turbo = false;
+  result.swingh = stdAc::swingh_t::kOff;
+  result.light = false;
+  result.filter = false;
+  result.econo = false;
+  result.quiet = false;
+  result.clean = false;
+  result.beep = false;
+  result.clock = -1;
```
What happened to Sleep?

danielegobbetti replied:
See the updated top comment (meanwhile I added sleep as unsupported with a TODO)

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,358 @@
+                              0xFF,
+                              0xFF, 0xFF,
+                              kYorkSwingVOff, kYorkSwingVOn,
+                              0xFF, 0xFF);
```
Per general comment:
```c++
  result += addBoolToString(getSwingV(), kSwingVStr);
```

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,123 @@
+    uint8_t Fan      :4;  // Fan speed: 1 low, 2 mid, 3 max, 8 auto
+    uint8_t Power    :4;  // main unit power: 1 on, 0 off
+    //byte 8
+    uint8_t Mode     :8;  // 1 heat, 2 cool, 3 dry, 4 fan, 8 auto
```
If those are the only possible values, then it is likely that `Mode` is only 4 bits.
so use:
```c++
    uint8_t Mode     :4;  // 1 heat, 2 cool, 3 dry, 4 fan, 8 auto
    uint8_t          :4;
```

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,123 @@
+    //byte 12
+    uint8_t          :8;  // unknown, normalyy 0x00, 0x08 when ontime set, 0x88 if both on and offtime set (could mean "auto power")
+    //byte 13
+    uint8_t SwingV   :8;  // 0 off, 1 on
```
If those are the only possible values, then it is likely that `SwingV` is a single bi.
so use:
```c++
    uint8_t SwingV   :1;  // 0 off, 1 on
    uint8_t          :7;
```

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,123 @@
+};
+
+// Constants
+const uint64_t kYorkKnownGoodState = 0x0810070240080318016000000000ECF5F2;  // Mode "Heat", Fan Speed "auto", Temp: 24, Power: on
```
This doesn't seem right. You've worked out it is a 136bit protocol. Which is larger than 64 bits.

danielegobbetti commented:
Many thanks for the thorough review and for your suggestions.

As for the OnTimer / OffTimer, additionally to the changes you suggested I also had to use getOnTimer() and getOffTimer in the Message description (it's obvious in hindsight).

I just resolved the comments where no comments where needed and added two more questions to the opening message.

crankyoldgit commented:
Still got a few Lint issues to address: (See: https://github.com/crankyoldgit/IRremoteESP8266/actions/runs/3134922966/jobs/5090021582)
i.e.
```
src/IRsend.h:845:  "protected:" should be preceded by a blank line  [whitespace/blank_line] [3]
src/ir_York.cpp:289:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_York.cpp:306:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+              kYorkBitMark, kYorkOneSpace,
+              kYorkBitMark, kYorkZeroSpace,
+              kYorkBitMark, kDefaultMessageGap,
+              data, kYorkStateLength,  // Bytes
```
change `kYorkStateLength` to `nbytes`

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+/// @return A boolean. True if it can decode it, false if it can't.
+bool IRrecv::decodeYork(decode_results *results, uint16_t offset,
+                        const uint16_t nbits, const bool strict) {
+  if (results->rawlen < 2 * nbits + kYorkOverhead - offset)
```
replace (and remove the constant entirely) `kYorkOverhead` with `kHeader + kFooter - 1`

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+  result += addTempToString(getTemp(), true);
+  result += addBoolToString(_.SwingV, kSwingVStr);
+  result += addLabeledString(minsToString(_.OnTimer), kOnTimerStr);
+  result += addLabeledString(minsToString(_.OffTimer), kOffTimerStr);
```
Change `_.OnTimer` to `getOnTimer()`.
Similar for the Off timer too.  You want the result to be in minutes, not 10's of minutes.

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,137 @@
+    // byte 11
+    uint8_t OnTimer  :8;  // Power on time: 10s of minutes from now
+    // byte 12
+    uint8_t          :8;  // unknown, normalyy 0x00, 0x08 when ontime set,
```
Typo: Normally

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,137 @@
+    // byte 12
+    uint8_t          :8;  // unknown, normalyy 0x00, 0x08 when ontime set,
+                          // 0x88 if both on and offtime set (could mean "auto
+                          // power")
```
Suggestion:
```c++
  // Byte 12
  uint8_t             :3;
  uint8_t OnTimerSet  :1;
  uint8_t             :3;
  uint8_t OffTimerSet :1;
```

That should allow you to easily test if they are set or not. Based on your observations of `0x00` all off, `0x08` On timer set. `0x80` Off timer set, `0x88` both set.

danielegobbetti replied:
My assumptions about those specific bytes were wrong, I'm back to the drawing board. See [here](https://docs.google.com/spreadsheets/d/1PsI8KSKzd8HQQeZUcCiCktq7l3xWww1AtdCdbUrWmfg/edit?usp=sharing)

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,137 @@
+const uint8_t kYorkAuto = 8;
+// Swing
+const uint8_t kYorkSwingVOn = 1;
+const uint8_t kYorkSwingVOff = 0;
```
You probably not need these two constants now.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+    case kYorkCool:
+    case kYorkHeat:
+    case kYorkDry:
+    case kYorkAuto:
```
Technically, you don't need the `case kYorkAuto:`, as it will be covered by the the `default:` clause.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+void IRYorkAc::send(const uint16_t repeat) {
+  _irsend.sendYork(getRaw(), kYorkBits, repeat);
+}
+#endif  // SEND_YORK
```
As you don't have the checksum calculation yet, you probably shouldn't provide a `send()` function/offering yet. People may assume it works. i.e. It sends the constructed state as a valid message.

danielegobbetti replied:
How is it best done? Do you prefer ``#define SEND_YORK false`` in IRremoteESP8266.h or rather an empty function?

crankyoldgit replied:
No. As you can still send a known-good state via sendYork().
Also, you've got `kYorkBits`, but `sendYork()` expects bytes. So `kYorkStateLength` etc.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRYorkAc::IRYorkAc(const uint16_t pin, const bool inverted,
```
I think you're missing a `IRYorkAc::stateReset()` function.
i.e. Call `setRaw(kYorkKnownGoodState);`

crankyoldgit replied:
You're still missing this. i.e. When the object is created, you want `stateReset()` to be called.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+const uint16_t kYorkOneSpace = 1778;
+const uint16_t kYorkZeroSpace = 579;
+const uint16_t kYorkFreq = 38000;  // Hz. (Guessing the most common frequency.)
+const uint16_t kYorkOverhead = 3;
```
See later review comment.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,340 @@
+using irutils::addModeToString;
+using irutils::addFanToString;
+using irutils::addTempToString;
+using irutils::addSwingVToString;
```
You can remove this line now.

crankyoldgit reviewed (commented):
Thanks for making the other changes .. getting closer now.

crankyoldgit commented:
> * It is not clear to me what the "sleep" supported feature (i.e. `result.sleep` in toCommon()) actually is: is it meant to be a "night/low noise mode" or rather "power the unit of after some time" mode?

Ha! Welcome to the world of AirConditioners and different brand/models/implementations.

To keep "Sleep" generic, it means what ever it means for that brand/model etc.
Some models have a "sleep" mode, i.e. "Run this A/C in a special sleep cycle" so it's treated like a `bool`. e.g. `-1` means off, `>=0` means on.
Some it's a turn off the A/C in X many minutes, in that case `-1`, or `0` is off. and anything larger than `0` is the number of minutes before it turns off.

So, it's what ever seems appropriate for the device.
TL;DR: It depends. ;-)


> * The on/offTimer do not seem to have a corresponding field in the state_t struct. Should these be mapped to some combination of clock / sleep?

No, we don't support On & Off timers in the "Common A/C" (`IRac` class) API. They are just too different to easily equate between brands & models. (e.g. On or Off, In X many mins turn off, Operate in sleep pattern for the next X mins, Turn on sleep mode at X many minutes past midnight (i.e. Set sleep mode at 10:15pm)

If the off timer is really needed by someone, via that interface, we might map it to `sleep` for them, _**if and only if**_. there is no sleep "mode" setting. i.e. An actual sleep setting/mode takes precidence over a hack to make an Off or On timer available via the `IRac` interface.

So, only set `sleep` to the number of mins, if the A/C supports a setable sleep duration etc. If it is just an on or off setting (i.e. `boolean`), then `-1` & `0` respectively.

**Edit:** Oh, ditto for `clock` too.

danielegobbetti commented:
> Still got a few Lint issues to address: (See: https://github.com/crankyoldgit/IRremoteESP8266/actions/runs/3134922966/jobs/5090021582) i.e.
> 
> ```
> src/IRsend.h:845:  "protected:" should be preceded by a blank line  [whitespace/blank_line] [3]
> ```

This one was introduced because you said so :-) :innocent: 
https://github.com/crankyoldgit/IRremoteESP8266/pull/1889#pullrequestreview-1121017161

In any case it's fixed now.

Edit: force pushed since I rebased to solve the conflicts.

danielegobbetti commented:
> > * It is not clear to me what the "sleep" supported feature (i.e. `result.sleep` in toCommon()) actually is: is it meant to be a "night/low noise mode" or rather "power the unit of after some time" mode?
> 
> Ha! Welcome to the world of AirConditioners and different brand/models/implementations.
> 
> To keep "Sleep" generic, it means what ever it means for that brand/model etc. Some models have a "sleep" mode, i.e. "Run this A/C in a special sleep cycle" so it's treated like a `bool`. e.g. `-1` means off, `>=0` means on. Some it's a turn off the A/C in X many minutes, in that case `-1`, or `0` is off. and anything larger than `0` is the number of minutes before it turns off.
> 
> So, it's what ever seems appropriate for the device. TL;DR: It depends. ;-)
> [...]
> So, only set `sleep` to the number of mins, if the A/C supports a setable sleep duration etc. If it is just an on or off setting (i.e. `boolean`), then `-1` & `0` respectively.
> 
> **Edit:** Oh, ditto for `clock` too.

Thanks, I did as you suggested and used the OffTimer (if set) as sleep value.

The remote control offer two buttons: 
one is labeled "sleep" and sets the off timer in the future with steps of 30 minutes (I guess for user convenience)
another one is labeled "stop" and is hidden behind a plastic flap, further it requires more button presses (one has to press "stop" than "timer up" for a number of times (steps is 10 minutes in this case) and finally "enter" to send the value to the AC unit.
Nevertheless the bits in the message are the same and the remote unsets (on its screen) the "sleep" value if anything is done with "stop" and vice versa.

Long story short: I believe it is correct to use sleep the way I now do, thanks a lot for the clarification.

Just to be extra sure I got your explanation correctly: there is no way to map in the IRAac class the status of "power on X minutes in the future", is this right?

crankyoldgit commented:
> Just to be extra sure I got your explanation correctly: there is no way to map in the IRAac class the status of "power on X minutes in the future", is this right?

Correct.

crankyoldgit commented:
Regarding the checksum(s). What have you tried.
Looking at your data, it doesn't look like a simple sum. Have you tried XORing?

danielegobbetti commented:
> Regarding the checksum(s). What have you tried. Looking at your data, it doesn't look like a simple sum. Have you tried XORing?

Yes I tried to make sense of it, in the second sheet there are two type of rows "MSG" and "XOR", where the latter is the xor of the 2 rows above.
Unfortunately I can hardly find a consistent correlation, I was hoping someone with more experience could find similarities with other protocols or anyway shed some light

NiKiZe commented on src/IRremoteESP8266.h:
```diff
@@ -931,6 +931,13 @@
+#endif  // DECODE_YORK
+#ifndef SEND_YORK
+#define SEND_YORK      _IR_ENABLE_DEFAULT_
+#endif  // SEND_YORK
```
please align the `_IR_ENABLE_DEFAULT_` to the protocols above

danielegobbetti replied:
I've put it at column 29, there are several different column alignments above...

NiKiZe commented on src/locale/defaults.h:
```diff
@@ -1057,6 +1057,9 @@ D_STR_INDIRECT " " D_STR_MODE
 #endif  // D_STR_VOLTAS
+#ifndef D_STR_YORK
+#define D_STR_YORK "YORK"
+#endif  // D_STR_YORK
```
These strings are sorted?, Y comes after X

NiKiZe reviewed (commented):
Just a few notes, needs to check more details, and having lack of time.

hoky24 commented:
> > Regarding the checksum(s). What have you tried. Looking at your data, it doesn't look like a simple sum. Have you tried XORing?
> 
> Yes I tried to make sense of it, in the second sheet there are two type of rows "MSG" and "XOR", where the latter is the xor of the 2 rows above. Unfortunately I can hardly find a consistent correlation, I was hoping someone with more experience could find similarities with other protocols or anyway shed some light

Coincidentally, I also have an old AC YORK, but with a GZ-12B-E1 remote. I just have no idea if it has the same IR codes. If you haven't figured out the checksum calculation yet, maybe I can help. 

I tried the online calculator https://www.scadacore.com/tools/programming-calculators/online-checksum-calculator/ for a quick reference. I entered the first three codes from the https://docs.google.com/spreadsheets/d/1PsI8KSKzd8HQQeZUcCiCktq7l3xWww1AtdCdbUrWmfg/edit#gid=1789386554 table without the last three bytes. 

Here is the result for the third code 0810070240080318017800000000

![obrazek](https://user-images.githubusercontent.com/3232374/223772708-00d32993-0a8d-4644-aba7-eb7dc1a2ea9d.png)

CRC-16-IBM (Bisync, Modbus, USB, ANSI X3.28, many others; also known as CRC-16 and CRC-16-ANSI), Big Endian (ABCD), Reversed 0xA001: D5 F0

danielegobbetti commented:
> 



> > > Regarding the checksum(s). What have you tried. Looking at your data, it doesn't look like a simple sum. Have you tried XORing?
> > 
> > 
> > Yes I tried to make sense of it, in the second sheet there are two type of rows "MSG" and "XOR", where the latter is the xor of the 2 rows above. Unfortunately I can hardly find a consistent correlation, I was hoping someone with more experience could find similarities with other protocols or anyway shed some light
> 
> Coincidentally, I also have an old AC YORK, but with a GZ-12B-E1 remote. I just have no idea if it has the same IR codes. If you haven't figured out the checksum calculation yet, maybe I can help.
> 
> I tried the online calculator https://www.scadacore.com/tools/programming-calculators/online-checksum-calculator/ for a quick reference. I entered the first three codes from the https://docs.google.com/spreadsheets/d/1PsI8KSKzd8HQQeZUcCiCktq7l3xWww1AtdCdbUrWmfg/edit#gid=1789386554 table without the last three bytes.
> 
> Here is the result for the third code 0810070240080318017800000000
> 
> ![obrazek](https://user-images.githubusercontent.com/3232374/223772708-00d32993-0a8d-4644-aba7-eb7dc1a2ea9d.png)
> 
> CRC-16-IBM (Bisync, Modbus, USB, ANSI X3.28, many others; also known as CRC-16 and CRC-16-ANSI), Big Endian (ABCD), Reversed 0xA001: D5 F0

Wow you have no idea how much time I lost on this. Thank you!

Edit: to clarify it looks like the checksum is calculated with the message truncated (-3 bytes), then recomposed like this:
truncated_message + "EC" + checksum

danielegobbetti commented:
Confirmed for all the codes I have on record:

```
$ ./reveng  -w 16 -s $(cat ../codes_without_EC.csv  | paste -sd" ")
width=16  poly=0x8005  init=0x0000  refin=true  refout=true  xorout=0x0000  check=0xbb3d  residue=0x0000  name="CRC-16/ARC"
```

danielegobbetti commented:
@crankyoldgit @NiKiZe 

I removed the WIP: prefix as I believe it might be ready to be merged now. I added checksum calculation and a few tests.
The sleep vs scheduled power on flags could be worked on later in my opinion, the basics should be working.

Regards,
Daniele

crankyoldgit commented:
@danielegobbetti Thanks for the heads up. I'll take a look now.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+bool IRrecv::decodeYork(decode_results *results, uint16_t offset,
+                        const uint16_t nbits, const bool strict) {
+  if (results->rawlen < 2 * nbits + (kHeader + kFooter - 1) - offset)
+    return false;  // Too short a message to match.
```
As this is all matched in a single `matchGeneric()` call, you can remove the length test, as `matchGeneric()` will do it.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRYorkAc::IRYorkAc(const uint16_t pin, const bool inverted,
+                         const bool use_modulation)
+      : _irsend(pin, inverted, use_modulation) { }
```
e.g.
```c++
      : _irsend(pin, inverted, use_modulation) { stateReset(); }
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+    case kYorkDry:
+      _.Fan = kYorkFanLow;
+      break;
+    case kYorkFan:
+      _.Fan = std::min(speed, kYorkFanHigh);
+      break;
```
You can remove this clause, it is redundant.

danielegobbetti replied:
this cannot be removed. In "Fan" mode the "Auto" speed is not available, hence it is capped to "High" in this case.

crankyoldgit replied:
Yep. I missed that.

danielegobbetti replied:
I was about to miss it as well, but fortunately the comment rang a bell :-)

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+      _.Fan = kYorkFanAuto;
+      break;
+    case kYorkCool:
+    case kYorkHeat:
```
You can remove these two cases, they are redundant.
i.e.
 Cool, Heat, & Fan modes will be caught by the `default:` clause.

danielegobbetti replied:
true with the exception of "Fan" mode, see above

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+void IRYorkAc::setTemp(const uint8_t degrees) {
+  uint8_t temp = std::max(kYorkMinTemp, degrees);
+  temp = std::min(kYorkMaxTemp, temp);
+  _.Temp = temp;
```
Replace with:
```c++
  _.Temp = std::min(kYorkMaxTemp, temp);
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+/// CRC16-16 (a.k.a. CRC-16-IBM)
+void IRYorkAc::calcChecksum() {
+int j;
+unsigned char length = 14;
```
please don't use `char`, use `uint8_t` instead please.
e.g.
```c++
  uint8_t length = 14;
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+int j;
+unsigned char length = 14;
+uint16_t reg_crc = 0x0000;
+unsigned char* data = _.raw;
```
ditto:
```c++
  uint8_t* data = _.raw;
```

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+
+/// CRC16-16 (a.k.a. CRC-16-IBM)
+void IRYorkAc::calcChecksum() {
+int j;
```
`j` is a terrible variable name. Use something useful/self-documenting. e.g.
```c++
uint16_t index;
```
etc.
In fact it can be delared later in the scope of the `for` loop later. See below.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+unsigned char* data = _.raw;
+while(length--) {
+  reg_crc ^= *data++;
+    for (j = 0; j < 8; j++) {
```
In fact, as `j`/`index` isn't used outside the loop, you can declare it here.
e.g.
```c++
    for (u_int16_t index = 0; index < 8; index++) {
```

danielegobbetti replied:
not sure about the difference between ``uint16_t`` and ``u_int16_t`` but I used the latter as per your comment.

crankyoldgit replied:
`u_int16_t` was a typo. Use `uint16_t`

crankyoldgit commented on src/ir_York.h:
```diff
@@ -0,0 +1,136 @@
+/// @brief Support for the York AC protocol (remote GRYLH2A)
+
+// Supports:
+//   Brand: York,  Model: MHH07P17
```
Specify it's an AC. (just a nitpick)
Is there a remote model number too? ;-) You should added that too.
e.g.
```c++
//   Brand: York,  Model: MHH07P17 A/C
//   Brand: York,  Model: GRYLH2A remote
```

crankyoldgit commented on test/ir_York_test.cpp:
```diff
@@ -0,0 +1,95 @@
+
+// General housekeeping
+TEST(TestYork, Housekeeping) {
+  ASSERT_EQ("YORK", typeToString(decode_type_t::YORK));
```
Also add a test for the other direction.
e.g.
```c++
  ASSERT_EQ(decode_type_t::YORK, strToDecodeType("YORK"));
```
You may have missed that code. Not sure.

crankyoldgit commented on test/ir_York_test.cpp:
```diff
@@ -0,0 +1,95 @@
+  EXPECT_STATE_EQ(kYorkKnownGoodState, irsend.capture.state,
+                  irsend.capture.bits);
+  EXPECT_EQ(
+      ", Power: On, Mode: 1 (Heat), Fan: 8 (Auto), Temp: 24C, Swing(V): Off"
```
The output should NOT start with ", ", that means you're missing a flag on the first output in the `::toString()` function.

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,362 @@
+String IRYorkAc::toString(void) const {
+  String result = "";
+  result.reserve(70);  // Reserve some heap for the string to reduce fragging.
+  result += addBoolToString(_.Power, kPowerStr);
```
First output will have ", " prepended unless you tell it not to do it.
e.g.
```c++
  result += addBoolToString(_.Power, kPowerStr, false);
```
See: https://crankyoldgit.github.io/IRremoteESP8266/doxygen/html/namespaceirutils.html#a12ba9cf1830a886649a80c3cc5fdce2b

crankyoldgit commented on test/ir_York_test.cpp:
```diff
@@ -0,0 +1,95 @@
+TEST(TestYorkClass, SetAndGetTemp) {
+  IRYorkAc ac(kGpioUnused);
+  ac.stateReset();
+  ac.setTemp(23);
```
You should check it wasn't `23` already.
e.g.
```c++
  EXPECT_NE(23, ac.getTemp());
  ac.setTemp(23);
  EXPECT_EQ(23, ac.getTemp());
```

crankyoldgit commented on test/ir_York_test.cpp:
You should probably have a lot more tests, but it's a good start for a new person. [No need to fix]

crankyoldgit reviewed (commented):
Looking good. Sorry for all the nitpicks. Nearly there.

danielegobbetti committed:
Address code review

danielegobbetti commented:
Thanks again for the prompt review. I truly deeply hope that now the current PR could be accepted because I have the impression that the part of the work that needs having a real device to test against is vanishing and my inexperience with c++ is frustrating for both sides.

In other words: I am available to test if everything still works after changing parts of the code against my A/C unit but I'm short of time (and knowledge) to dedicate a substantial effort to this PR if it's not yet deemed ready to be merged. Please don't get me wrong: I really appreciated your patience and guidance toward making my code decent enough :-)

crankyoldgit commented on src/ir_York.cpp:
```diff
@@ -0,0 +1,357 @@
+        reg_crc = reg_crc >>1;
+      }
+    }
+}
```
One last style nitpick. All of this should be indented correctly. i.e. These particular lines should be indented 2 spaces.

crankyoldgit reviewed (commented):
Looks good to me. Just the style nitpick to go.

danielegobbetti committed:
fix indentation and incorrect typing

danielegobbetti commented:
Should be ok now. Many thanks again.

crankyoldgit committed:
Fix linter issue

src/ir_York.cpp:290:  Missing space before ( in while(  [whitespace/parens] [5]

crankyoldgit reviewed (approved):
LGTM!
